/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
syntax = "proto3";

package profiler.proto;
option java_package = "com.android.tools.profiler.proto";
option java_outer_classname = "Common";

enum AppId {
  // Enum with default value required by proto
  UNSPECIFIED = 0;
  // Indicates that a feature is not tied to any single app
  ANY = -1;
}

// TODO: Add more essential fields.
message CommonData {
  int32 process_id = 1;
  Session session = 2;
  int64 end_timestamp = 3;
}

message Device {
  enum State {
    UNSPECIFIED = 0;
    ONLINE = 1;
    OFFLINE = 2; // e.g. emulator that is turned off
    DISCONNECTED = 3; // e.g. unplugged
  }
  string manufacturer = 1; // e.g. 'Google', 'Motorola'
  string model = 2; // e.g. 'Nexus 5', 'Pixel XL'
  string serial = 3; // unique ID for this device
  string version = 4; // e.g. '5.1.1', '8.0.0'
  int32 api_level = 5; // e.g. 19, 25
  string codename = 10; // Identifies preview versions by API letter (e.g. 'O'), or '' for release
  // |feature_level| will either be the same as |api_level| or |api_level + 1| if the current
  // device is a preview version. Code should check against |feature_level|, not |api_level|, to
  // see if a feature is available in the framework or not. If |feature_level != api_level|, then
  // |codename| will also be set.
  int32 feature_level = 6;
  string boot_id = 7; // unique ID regenerated each boot time
  bool is_emulator = 8;
  State state = 9;
}

message Process {
  enum State {
    UNSPECIFIED = 0;
    ALIVE = 1;
    DEAD = 2;
  }
  // Full name of the Android application/service
  string name = 1;
  // App's PID. Note that this is actually an int16, but gRPC only provides int32.
  int32 pid = 2;
  State state = 3;
  // The device time when this Process was first detected by the profiler.
  int64 start_timestamp_ns = 4;
  // e.g. 'arm', 'arm86', 'x86', 'x86_64'
  // Also see SdkConstans.CPU_ARCH_*
  string abi_cpu_arch = 5;
}

// A Session represents a range of profiling data for an app running on a
// device. While of course different apps or different devices will imply
// separate sessions, the same app on the same device can also be broken up
// into several sessions.
message Session {
  // ID uniquely identifying this session. This will be unique across devices.
  int64 session_id = 1;
  // A string which uniquely identifies this device.
  string device_serial = 2;
  // A string which uniquely identifies this particular device run. Restarting
  // the device will generate a new |boot_id|.
  string boot_id = 3;
  // The PID of the app. Note that this is actually an int16, but gRPC only
  // provides int32.
  int32 pid = 4;
  int64 start_timestamp = 5;
  // If LLONG_MAX, the session is ongoing and hasn't ended yet.
  int64 end_timestamp = 6;
}